ql/src/java/org/apache/hadoop/hive/ql/parse/ReturnPathManager.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.hive.ql.parse;
+
+import org.apache.hadoop.hive.conf.HiveConf;
+
+/**
+ * This is a temporary class for the time while we are in transition to use return path only.
+ * It's purpose is to decide if Hive should use return path.
+ */
+public final class ReturnPathManager {
+  private ReturnPathManager() {
+    throw new UnsupportedOperationException("ReturnPathManager should not be instantiated");
+  }
+
+  private static ThreadLocal<Boolean> useReturnPath = new ThreadLocal<>();
+
+  public static void init(HiveConf conf, ASTNode root) {
+    boolean use = shouldUseReturnPath(conf, root);
+    useReturnPath.set(use);
+  }
+
+  public static boolean shouldUse() {
+    return useReturnPath.get();
+  }
+
+  private static boolean shouldUseReturnPath(HiveConf conf, ASTNode root) {
```

Comment:
I believe it is better to make this decision in terms of `RelNode` plan rather than `ASTNode` since the optimizer may introduce some arbitrary complex operators and expressions. You can implement a custom `RelVisitorImpl` that traverses the tree and returns whether any operator/expression is supported or not. The visitor would be executed before the block that applies the rules to aid in translation from Calcite tree to Hive tree. I also think we can potentially revisit the decision on checking that return path boolean when we create the HiveTableScan operator.

Reply:
During the generating of the RelNode tree the CalcitePlannerAction uses several times the value of HIVE_CBO_RETPATH_HIVEOP, so I thought that it must be decided if for the current command processing we want to use return path or not in advance, so to have the proper value replacing the original boolean property consistently. Isn't that the case?

Reply:
If possible, let's remove those usages as we are discussing above... It is easier to do this check on the Calcite plan since optimizations may introduce/remove operators and expressions that are supported/unsupported.

Reply:
Ok, so I know now how to remove those usages from HiveTableScan creation, but what about the rest of the usages? CalcitePlannerAction.applyPostJoinOrderingTransform and CalcitePlannerAction.genTableLogicalPlan also relies on the value, can those be replaced with some hard coded value as well?

Reply:
The call in `genTableLogicalPlan` seems to be another unsupported feature? But it is unsupported for CBO in general... Hence I am confused about what the condition is actually doing.
Additionally, the validation code described above should be called just before `CalcitePlannerAction.applyPostJoinOrderingTransform`, since that method applies optimizations that are specific for RP; until that point, I expect that RP and AST path are exactly the same. Hence, if plan is not supported through RP, we can skip that block and return through AST.